openni/plugin.cpp

```diff
@@ -0,0 +1,160 @@
+        } else {
+            return skip_option::skip_and_yield;
+        }
+    }
+
+    virtual void _p_one_iteration() override {
+        RAC_ERRNO_MSG("openni at start of _p_one_iteration");
+
+        // read cam data
+        color.readFrame(&colorFrame);
+        depth.readFrame(&depthFrame);
+        // get timestamp
+        assert(colorFrame.getTimestamp() != depthFrame.getTimestamp());
+
+        // convert to cv format
+        cv::Mat colorMat;
+        colorMat.create(colorFrame.getHeight(), colorFrame.getWidth(), CV_8UC3);
+        const openni::RGB888Pixel* colorBuffer = (const openni::RGB888Pixel*) colorFrame.getData();
+        memcpy(colorMat.data, colorBuffer, 3 * colorFrame.getHeight() * colorFrame.getWidth() * sizeof(uint8_t));
+        cv::cvtColor(colorMat, colorMat, cv::COLOR_BGR2BGRA);
+        colorMat.convertTo(colorMat, CV_8UC4);
+
+        cv::Mat depthMat;
+        depthMat.create(depthFrame.getHeight(), depthFrame.getWidth(), CV_16UC1);
+        const openni::DepthPixel* depthBuffer = (const openni::DepthPixel*) depthFrame.getData();
+        memcpy(depthMat.data, depthBuffer, depthFrame.getHeight() * depthFrame.getWidth() * sizeof(uint16_t));
+
+        assert(cam_time);
+        if (_m_first_time == 0) {
+            _m_first_time      = cam_time;
+            _m_first_real_time = _m_clock->now();
+        }
+        time_point cam_time_point{_m_first_real_time + std::chrono::nanoseconds(cam_time - _m_first_time)};
+        _m_rgb_depth.put(_m_rgb_depth.allocate(cam_time_point, colorMat, depthMat));
+
+        last_ts = cam_time;
+        RAC_ERRNO_MSG("openni at end of _p_one_iteration");
+    }
+
+    bool camera_initialize() {
+        // initialize openni
+        rc = openni::OpenNI::initialize();
+        if (rc != openni::STATUS_OK)
+            std::cout << "Initialize failed: " << openni::OpenNI::getExtendedError() << std::endl;
+
+        // open device
+        rc = device.open(openni::ANY_DEVICE);
+        if (rc != openni::STATUS_OK)
+            std::cout << "Device open failed: " << openni::OpenNI::getExtendedError() << std::endl;
+
+        /*_____________________________ DEPTH ___________________________*/
+        // create depth channel
+        rc = depth.create(device, openni::SENSOR_DEPTH);
+        if (rc != openni::STATUS_OK)
+            printf("Couldn't find depth stream:\n%s\n", openni::OpenNI::getExtendedError());
+
+#ifndef NDEBUG
+        // get depth options
+        const openni::SensorInfo*               depthInfo  = device.getSensorInfo(openni::SENSOR_DEPTH);
+        const openni::Array<openni::VideoMode>& modesDepth = depthInfo->getSupportedVideoModes();
+        for (int i = 0; i < modesDepth.getSize(); i++) {
+            printf("Depth Mode %i: %ix%i, %i fps, %i format\n", i, modesDepth[i].getResolutionX(),
+                   modesDepth[i].getResolutionY(), modesDepth[i].getFps(), modesDepth[i].getPixelFormat());
+        }
+        rc = depth.setVideoMode(modesDepth[4]);
```

Comment:
What are the modesDepth?
Is this the only mode we want to use for a particular OpenNI device?
Is this hardcoded No.4 applicable to all other OpenNI supported devices?

Reply:
Line 102-103 prints out all of the mode and resolution of Depth and the associate index, 4 is just the one that we use for ORB-SLAM + ITM. I will set it back to 0, and have a comment specifying what it does

Reply:
The hardcoded No.4 specifies that we wants to run depth mode at 640x480 res and 30 fps. I'm not sure if they applies to all OpenNI supported device, hence I have [this line](https://github.com/ILLIXR/ILLIXR/blob/8f40cf21d34bac14cf125c0758329becab95718c/openni/plugin.cpp#L102) print all the available mode out, which must then be set manually by the user. I will set them back to default.